File: iosApp/iosApp/Features/ComfyHub/ComfyHubBrowserViewModel.swift
import Foundation
import Shared

@MainActor
final class ComfyHubBrowserViewModelOwner: ObservableObject {
    let vm: ComfyHubBrowserViewModel
    private let store = ViewModelStore()

    @Published var workflows: [ComfyHubWorkflow] = []
    @Published var isLoading: Bool = true
    @Published var error: String?
    @Published var query: String = ""
    @Published var selectedCategory: ComfyHubCategory = .all

    init() {
        vm = KoinHelper.shared.createComfyHubBrowserViewModel()
        store.put(key: "ComfyHubBrowserViewModel", viewModel: vm)
    }

    deinit { store.clear() }

    func observeUiState() async {
        for await state in vm.uiState {
            workflows = state.workflows as? [ComfyHubWorkflow] ?? []
            isLoading = state.isLoading
            error = state.error
            query = state.query
            selectedCategory = state.selectedCategory
        }
    }

    func onQueryChanged(_ query: String) {
        self.query = query
        vm.onQueryChange(query: query)
    }
    func onCategorySelected(_ category: ComfyHubCategory) {
        selectedCategory = category
        vm.onCategorySelected(category: category)
    }
    func retry() { vm.retry() }
}

File: shared/src/commonMain/kotlin/com/riox432/civitdeck/presentation/comfyhub/ComfyHubBrowserViewModel.kt
package com.riox432.civitdeck.presentation.comfyhub

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.riox432.civitdeck.domain.model.ComfyHubCategory
import com.riox432.civitdeck.domain.model.ComfyHubSortOrder
import com.riox432.civitdeck.domain.model.ComfyHubWorkflow
import com.riox432.civitdeck.domain.util.UiLoadingState
import com.riox432.civitdeck.feature.comfyui.domain.usecase.SearchComfyHubWorkflowsUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class ComfyHubBrowserUiState(
    val workflows: List<ComfyHubWorkflow> = emptyList(),
    override val isLoading: Boolean = true,
    override val error: String? = null,
    val query: String = "",
    val selectedCategory: ComfyHubCategory = ComfyHubCategory.ALL,
    val sortOrder: ComfyHubSortOrder = ComfyHubSortOrder.MOST_DOWNLOADED,
) : UiLoadingState

class ComfyHubBrowserViewModel(
    private val searchWorkflows: SearchComfyHubWorkflowsUseCase,
) : ViewModel() {

    private val _uiState = MutableStateFlow(ComfyHubBrowserUiState())
    val uiState: StateFlow<ComfyHubBrowserUiState> = _uiState.asStateFlow()

    init {
        search()
    }

    fun onQueryChange(query: String) {
        _uiState.update { it.copy(query = query) }
        search()
    }

    fun onCategorySelected(category: ComfyHubCategory) {
        _uiState.update { it.copy(selectedCategory = category) }
        search()
    }

    fun onSortOrderChanged(sort: ComfyHubSortOrder) {
        _uiState.update { it.copy(sortOrder = sort) }
        search()
    }

    fun retry() {
        search()
    }

    @Suppress("TooGenericExceptionCaught")
    private fun search() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            try {
                val state = _uiState.value
                val results = searchWorkflows(
                    query = state.query,
                    category = state.selectedCategory,
                    sort = state.sortOrder,
                )
                _uiState.update { it.copy(workflows = results, isLoading = false) }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(error = e.message ?: "Search failed", isLoading = false)
                }
            }
        }
    }
}
